### PR TITLE
libmtd: add Memory Technology Device (MTD) Interface

### PR DESCRIPTION
**Pull request by HBuczynski**
## Description

Library for Memory Technology Device (MTD) Interface. This PR provides basic functionality of MTD device which has been tested in `jffs2` filesystem.

DTR-102, PD-216

## Motivation and Context

The library provides an abstraction layer for NOR and NAND flash memories used by filesystems and `mtd-utils` tools to manage raw operations on flash memories.

To provide appropriate interface for `mtd-utils`, additional structures have to be added (based on the linux interface):
 - [mtd-abi](https://github.com/torvalds/linux/blob/master/include/uapi/mtd/mtd-abi.h)
 - [mtd-user](https://github.com/torvalds/linux/blob/master/include/uapi/mtd/mtd-user.h)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (zynq7000).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (https://github.com/phoenix-rtos/libphoenix/pull/174).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by agkaminski — CHANGES_REQUESTED** [review]
MISRA related remarks applies to the whole PR

File: libmtd/include/mtd/mtd.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+/* Lock from writing and erasing a specific flash memory area */
+extern int mtd_lock(struct mtd_info *mtd, off_t ofs, uint64_t len);
+
+
+/* Unlock from writing and erasing a specific flash memory area */
+extern int mtd_unlock(struct mtd_info *mtd, off_t ofs, uint64_t len);
+
+
+/* Check locking bit of a specific flash memory area */
+extern int mtd_is_locked(struct mtd_info *mtd, off_t ofs, uint64_t len);
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply libmtd/include/mtd/mtd.h:158]: Yes, there are standard functions name used by `jffs2` and other tools.

File: libmtd/mtd.c [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+	if (instr->addr >= mtdInfo->size || instr->len > mtdInfo->size - instr->addr)
+		return -EINVAL;
+
+	if (!(mtdInfo->flags & MTD_WRITEABLE))
+		return -EROFS;
+
+	instr->fail_addr = MTD_FAIL_ADDR_UNKNOWN;
+	instr->state = MTD_ERASING;
+	if (!instr->len) {
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply libmtd/mtd.c:38]: Which rule define this?
`!` is a logical operand which returns `true` or `false`. The `Rule 14.4` says:

> The controlling expression of an if statement and the controlling
expression of an iteration- statement shall have essentially Boolean type

Moreover, I have seen in the other PRs that such statements were accepted.

File: libmtd/mtd.c [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+#include <stdlib.h>
+
+
+int mtd_erase(struct mtd_info *mtdInfo, struct erase_info *instr)
+{
+	int res;
+	storage_mtd_t *mtd = mtdInfo->storage->dev->mtd;
+
+	if (mtd->ops->erase == NULL)
+		return -EOPNOTSUPP;
```
[the remark this answers is not in the document]
Comment (HBuczynski) [reply libmtd/mtd.c]: Thanks for comment.
I have changed the code in the whole PR to be consistent with MISRA rules.

File: libmtd/mtd.c [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+		instr->fail_addr = instr->addr;
+		instr->state = MTD_ERASE_FAILED;
+	}
+	else {
+		instr->state = MTD_ERASE_DONE;
+	}
+
+	if (instr->callback != NULL) {
+		instr->callback(instr);
+	}
```
Comment (agkaminski) [inline libmtd/mtd.c:62]: One exit path and code repetition less
```suggestion
	if (instr->len == 0) {
		instr->state = MTD_ERASE_DONE;
	}
	else {
		instr->state = MTD_ERASE_PENDING;
		res = mtd->ops->erase(mtdInfo->storage, instr->addr, instr->len);
		if (res < 0) {
			instr->fail_addr = instr->addr;
			instr->state = MTD_ERASE_FAILED;
		}
		else {
			instr->state = MTD_ERASE_DONE;
		}
	}

	if (instr->callback != NULL) {
		instr->callback(instr);
	}
```

File: libmtd/mtd.c [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+
+	if (from < 0 || from >= mtdInfo->size || len > (mtdInfo->size - from)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return EOK;
+	}
+
+	return mtd->ops->point(mtdInfo->storage, from, len, retlen, virt, phys);
```
Comment (agkaminski) [inline libmtd/mtd.c:91]: Avoiding multiple exit paths
```suggestion
	if (mtd->ops->point == NULL) {
		ret = -EOPNOTSUPP;
	}
	else if (from < 0 || from >= mtdInfo->size || len > (mtdInfo->size - from)) {
		ret = -EINVAL;
	}
	else if (len == 0) {
		ret = EOK;
	}
	else {
		ret = mtd->ops->point(mtdInfo->storage, from, len, retlen, virt, phys);
	}
	
	return ret;
```

Comment (agkaminski) [reply libmtd/mtd.c:91]: Most other cases can be rearranged in a similar fasion

File: libmtd/mtd.c (hunk 4 above)
Comment (HBuczynski) [reply libmtd/mtd.c]: Applied Your suggestion.

File: libmtd/mtd.c [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+void *mtd_kmalloc_up_to(const struct mtd_info *mtdInfo, size_t *size)
+{
+	void *ret;
+
+	if (*size < mtdInfo->writesize) {
+		*size = mtdInfo->writesize;
+	}
+
+	if (mtdInfo->writesize < _PAGE_SIZE) {
+		ret = mmap(NULL, *size, PROT_READ | PROT_WRITE, MAP_ANONYMOUS, OID_CONTIGUOUS, 0);
```
Comment (agkaminski) [inline libmtd/mtd.c:513]: Won't work for `*size != n * _PAGE_SIZE`. We need to align `*size` both to `mtdInfo->writesize` and `_PAGE_SIZE`.

File: libmtd/mtd.c [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+
+	addr = mmap(NULL, size, PROT_WRITE | PROT_READ, MAP_ANONYMOUS, OID_CONTIGUOUS, 0);
+	if (addr == MAP_FAILED) {
+		return NULL;
+	}
+
+	pa = va2pa(addr);
+	if ((pa % alignment) == 0) {
+		return addr;
+	}
```
Comment (agkaminski) [inline libmtd/mtd.c:479]: Do we need aligned physical or virtual addresses?

File: libmtd/mtd.c [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+
+void *mtd_kmalloc_up_to(const struct mtd_info *mtdInfo, size_t *size)
+{
+	void *ret;
+
+	if (*size < mtdInfo->writesize) {
+		*size = mtdInfo->writesize;
+	}
+
+	if (mtdInfo->writesize < _PAGE_SIZE) {
```
Comment (agkaminski) [inline libmtd/mtd.c:512]:
```suggestion
	if (mtdInfo->writesize <= _PAGE_SIZE) {
```

**Review by agkaminski — APPROVED** [review]

**Pull request merged**